cmd/snap-confine/snap-confine.c

```diff
@@ -257,6 +264,8 @@ int main(int argc, char **argv)
 
 			sc_unlock(snap_name, snap_lock_fd);
 
+			sc_setup_user_mounts(snap_update_ns_fd, snap_name);
```

Comment:
I think that it doens't have to yet but I agree it should probably be inside some form of "I'm looking at this NS" lock eventually.

Reply:
With the current design, `sc_setup_user_mounts` is doing the following:

* unshare the mount namespace
* switch all mounts to MS_SLAVE mode, so we receive upstream changes but not vice versa
* apply the contents of the `.user-fstab` file.

If the change in the second step (which is achieved with a single system call) is not atomic, I guess it would be possible for some mounts in the ephemeral namespace to remain in MS_SHARED mode.  This would have the possibility of leaking a user mount to the persistent namespace if its parent directory was still in MS_SHARED mode.

With that said, the agreement is to only use the current form of this patch to support xdg-document-portal, which mounts under $XDG_RUNTIME_DIR.  No snapd interfaces introduce mounts between the root and this directory, so this wouldn't happen in practice.